domain/resource/state/resource.go

```diff
@@ -407,198 +421,551 @@ WHERE uuid = $resourceIdentity.uuid`,
 		return resource.Resource{}, errors.Capture(err)
 	}
 
-	return resourceOutput.toResource(), nil
+	return resourceOutput.toResource()
 }
 
-// SetResource adds the resource to blob storage and updates the metadata.
-func (st *State) SetResource(
+// StoreResource records a stored resource along with who retrieved it.
+//   - [resourceerrors.ResourceNotFound] if the resource UUID cannot be found.
+//   - [resourceerrors.StoredResourceNotFound] if the stored resource at the
+//     storageID cannot be found.
+//   - [resourceerrors.ResourceAlreadyStored] if the resource is already
+//     associated with a stored resource blob.
+//   - [resourceerrors.RetrievedByTypeNotValid] if the retrieved by type is
+//     invalid.
+func (st *State) RecordStoredResource(
 	ctx context.Context,
-	config resource.SetResourceArgs,
-) (resource.Resource, error) {
-	return resource.Resource{}, nil
+	resourceUUID coreresource.UUID,
+	storageID coreresourcestore.ID,
+	retrievedBy string,
+	retrievedByType resource.RetrievedByType,
+	incrementCharmModifiedVersion bool,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		kind, err := st.getResourceType(ctx, tx, resourceUUID)
```

Comment:
Given that the resource type isn't subject to change, I would put this retrieval in the service layer.

Reply:
Fair point, have moved it there.